src/cargo/util/toml/mod.rs

```diff
@@ -2154,3 +2225,28 @@ impl ResolveToPath for ConfigRelativePath {
         self.resolve_path(c)
     }
 }
+
+fn translate_position(input: &str, index: usize) -> (usize, usize) {
+    if input.is_empty() {
+        return (0, index);
+    }
+
+    let safe_index = index.min(input.len() - 1);
+    let column_offset = index - safe_index;
+
+    let nl = input[0..safe_index]
+        .chars()
+        .rev()
+        .enumerate()
+        .find(|(_, c)| c == &'\n')
+        .map(|(nl, _)| safe_index - nl - 1);
+    let line_start = match nl {
+        Some(nl) => nl + 1,
+        None => 0,
+    };
+    let line = input[0..line_start].chars().filter(|c| c == &'\n').count();
```

Comment:
nit: If you can't use something like `.lines()` then I'd recommend doing `as_bytes()` instead of `chars()` as newline counting can work on the byte level and `chars()` is slow.